README.md
# climy
Do CLI applications easily.


## Instalation
```shell
$ pip install climy
```

### Basic Usage
```python
from climy import Application, Option, Command


def env_handler(comm: Command):
    print('do every thing with env')
    print(comm)
    print(comm.vars)

    
env_command = Command(
    name='env',
    description='Create a perfect environ to application',
    handler=env_handler
)

app = Application(
    name='acme',
    title='ACME App',
    description='A teste application',
    version='0.1.0',
    option_default_separator=Option.Separators.EQUAL
)
app.add_command(env_command)
app.add_option('host', 'The server allow host IP number.', short='t', var_type='str')
app.add_option('port', 'Service port.', short='p', var_type='int')
app.add_option('path', 'Execution code path.', var_type='str', var_name='path')
app.run()
```
```shell
$ acmesrv --help
ACME App (version 0.1.0)

Usage:
  acme <command> [options] [arguments]

Options:
  -h, --help                Print this help.
  -t, --host=[HOST]         The server allow host IP number.
  -p, --port=[PORT]         Service port.
  --path=[PATH]             Execution code path.

Commands:
  env                       Create a perfect environ to application
```

## Components

### Tables
To print tables quickly, you just need to define a column list with the column title, size and align. 
Then, preparate a data tuple list with the values in the same columns orders. 
Below, an example how to print tables.
```python
from climy.console import ctable


columns = ['ID:10>', 'Name:50', 'Birth:15^', 'Weight:10>', 'Hight:10']
data = [
    (1, 'Mark White', '26/12/1977', 1.75, 80.0),
    (2, 'Eva Wood', '30/11/1983', 1.59, 50.0),
    (3, 'John Apple', '26/06/1917', 1.20, 30.0),
    (4, 'Linda Jansen', '09/05/1961', 1.63, 61.0)
]
ctable(data, columns=columns)
```
And the results:
```
┌──────────┬──────────────────────────────────────────────────┬───────────────┬──────────┬──────────┐
│       ID │ Name                                             │     Birth     │   Weight │ Hight    │
├──────────┼──────────────────────────────────────────────────┼───────────────┼──────────┼──────────┤
│        1 │ Mark White                                       │   26/12/1977  │     1.75 │ 80.0     │
│        2 │ Eva Wood                                         │   30/11/1983  │     1.59 │ 50.0     │
│        3 │ John Apple                                       │   26/06/1917  │      1.2 │ 30.0     │
│        4 │ Linda Jansen                                     │   09/05/1961  │     1.63 │ 61.0     │
└──────────┴──────────────────────────────────────────────────┴───────────────┴──────────┴──────────┘
```

### Progress Bar
You can print progresses bars by using cprogress function.

```python
from climy.console import cprogress


cprogress(size=50, max=300.0, value=20.0, color='green')

# ━━━───────────────────────────────────────────────
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]


## [0.2.1]
### Fixed
- Fixed error on sprint method. It was print itself.


## [0.2.0] - 2023-07-07
### Added
- New module console with styled print options and some components like ctable and cprogress.
- Changelog file to monitory versions changes. 
- New Handler class to manage handler commands executions.


## [0.1.1] - 2023-06-30
### Fixed
- Fix an error when a subcommand options was passed.

## [0.1.0] - 2032-06-29
### Added
- Application class to create new cli applications.
- Auto help generation based on commands and options added on application.
- Commands can have subcommands with auto help generation and options.


## [0.0.0] - 2023-06-02
### Added
- Basic project structure.

[Unreleased]: https://github.com/manasseslima/climy/compare/v0.2.0...HEAD
[0.1.0]: https://github.com/manasseslima/climy/releases/tag/v0.1.0

tests/test_console.py
from climy.console import ctable, cprogress

def test_print_table():


    columns = ['ID:10>', 'Name:0', 'Birth:15^', 'Weight:10>', 'Hight:10>', 'Gender:10^']
    data = [
        (1, 'Mark White', '26/12/1977', 1.75, 80.0, 'Male'),
        (2, 'Eva Wood', '30/11/1983', 1.59, 50.0, 'Female'),
        (3, 'John Apple', '26/06/1917', 1.20, 30.0, 'Male'),
        (4, 'Linda Jansen', '09/05/1961', 1.63, 61.0, 'Female')
    ]
    ctable(data, columns=columns)
    assert True


def test_print_progress():
    cprogress(size=50, max=300.0, value=20.0, color='green')


climy/console.py
import shutil
import time

CINIT = '\033[{}m'
CEND = '\033[0m'
CBLOCK = '\033[{}m{}\033[0m'

CBOLD = '1'
CDIM = '2'
CITALIC = '3'
CURL = '4'
CSBLINK = '5'
CFBLINK = '6'
CSELECTED = '7'

CBLACK = '30'
CRED = '31'
CGREEN = '32'
CYELLOW = '33'
CBLUE = '34'
CVIOLET = '35'
CCYAN = '36'
CGREY = '37'
CWHITE = '38'

CBLACKBG = '40'
CREDBG = '41'
CGREENBG = '42'
CYELLOWBG = '43'
CBLUEBG = '44'
CVIOLETBG = '45'
CCYANBG = '46'
CGREYBG = '47'
CWHITEBG = '48'

CBLACKH = '90'
CREDH = '91'
CGREENH = '92'
CYELLOWH = '93'
CBLUEH = '94'
CVIOLETH = '95'
CCYANH = '96'
CGREYH = '97'
CWHITEH = '98'

CBLACKBGH = '100'
CREDBGH = '101'
CGREENBGH = '102'
CYELLOWBGH = '103'
CBLUEBGH = '104'
CVIOLETBGH = '105'
CCYANBGH = '106'
CGREYBGH = '107'
CWHITE = '108'

PGBLOCK = '\u2588'

HSOLID = '\u2500'
VSOLID = '\u2502'

TLSOLID = '\u250C'
TRSOLID = '\u2510'
TDSOLID = '\u252C'

BLSOLID = '\u2514'
BRSOLID = '\u2518'
BUSOLID = '\u2534'

CLSOLID = '\u251C'
CSOLID = '\u253C'
CRSOLID = '\u2524'

cstyles = {
    'bold': '1',
    'dim': '2',
    'italic': '3',
    'underline': '4',
    'black': CBLACK,
    'red': CRED,
    'green': CGREEN,
    'yellow': CYELLOW,
    'blue': CBLUE,
    'violet': CVIOLET,
    'cyan': CCYAN,
    'grey': CGREY,
    'white': CWHITE,
    'bgblack': CBLACKBG,
    'bgred': CREDBG,
    'bggreen': CGREENBG,
    'bgyellow': CYELLOWBG,
    'bgblue': CBLUEBG,
    'bgviolet': CVIOLETBG,
    'bgcyan': CCYANBG,
    'bggrey': CGREYBG,
    'bgwhite': CWHITE,
    'hblack': CBLACKH,
    'hred': CREDH,
    'hgreen': CGREENH,
    'hyellow': CYELLOWH,
    'hblue': CBLUEH,
    'hviolet': CVIOLETH,
    'hcyan': CCYANH,
    'hgrey': CGREYH,
    'hwhite': CWHITEH,
    'hbgblack': CBLACKBGH,
    'hbgred': CREDBGH,
    'hbggreen': CGREENBGH,
    'hbgyellow': CYELLOWBGH,
    'hbgblue': CBLUEBGH,
    'hbgviolet': CVIOLETBGH,
    'hbgcyan': CCYANBGH,
    'hbggrey': CGREYBGH,
    'hbgwhite': CWHITE,
}


def styled_text(text: str, style: str):
    splt = style.split(' ')
    ccode = ';'.join([cstyles.get(item, '39') for item in splt])
    return CBLOCK.format(ccode, text)


def sprint(text: str, style='', align='<', end='\n'):
    stext = styled_text(text, style=style)
    if align == '>':
        tml_columns = shutil.get_terminal_size().columns
        size = len(text)
        stext = ' ' * (tml_columns - size) + stext
    elif align == '^':
        tml_columns = shutil.get_terminal_size().columns
        stext = stext.center(tml_columns)
    print(stext, end=end)


def cmove(x, y):
    print('\033[{x};{y}HTESTE'.format(x=x, y=y))


class CTable:
    def __init__(self, columns: list[str]):
        self.columns = columns


def ctable(
        data: list,
        columns: list[str],
        style='',
        end_last=True,
        print_header=True
):
    tml_columns = shutil.get_terminal_size().columns
    tml_rest = tml_columns
    tcols = {}
    thline = f'{TLSOLID}'
    bhline = f'{CLSOLID}'
    iline = f'{CLSOLID}'
    bline = f'{BLSOLID}'
    htext = f'{VSOLID}'
    for idx, col in enumerate(columns):
        splt = col.split(':')
        name = splt[0]
        if len(splt) > 1:
            align = splt[1][-1] if splt[1][-1] in ['<', '^', '>'] else '<'
            size = int(splt[1][:-1] if splt[1][-1] in ['<', '^', '>'] else splt[1])
            tml_rest -= size
        else:
            align = '<'
            size = 0
        cl = {
            'name': name,
            'align': align,
            'size': size,
            'last': idx == len(columns) - 1
        }
        tcols[idx] = cl
    for idx, col in tcols.items():
        if col['size'] == 0:
            col['size'] = tml_rest - (len(tcols) + 1)
        size = col['size']
        last = col['last']
        name = col['name']
        align = col['align']
        thline += HSOLID * size + (TRSOLID if last else TDSOLID)
        bhline += HSOLID * size + (CRSOLID if last else CSOLID)
        iline += HSOLID * size + (CRSOLID if last else CSOLID)
        bline += HSOLID * size + (BRSOLID if last else BUSOLID)
        match align:
            case '<':
                htext += ' ' + name.ljust(size - 1) + VSOLID
            case '^':
                htext += name.center(size) + VSOLID
            case '>':
                htext += name.rjust(size - 1) + ' ' + VSOLID
    if print_header:
        sprint(thline, style=style)
        sprint(htext, style=style)
        sprint(bhline, style=style)
    for item in data:
        dline = VSOLID
        for idx, val in enumerate(item):
            col = tcols[idx]
            match col['align']:
                case '<':
                    dline += ' ' + str(val).ljust(col['size'] - 1) + VSOLID
                case '^':
                    dline += str(val).center(col['size']) + VSOLID
                case '>':
                    dline += str(val).rjust(col['size'] - 1) + ' ' + VSOLID
        sprint(dline, style=style)
    if end_last:
        sprint(bline, style=style)
    else:
        sprint(iline, style=style)


def cprogress(
        size: int,
        max=0.0,
        value=0.0,
        color='grey',
        before: str = '',
        after: str = ''
):
    fator = size / max
    maked_size = int(value * fator)
    rest_size = size - maked_size
    maked = styled_text(('\u2501' * maked_size), style=f'{color} bold')
    rest = styled_text(('\u2500' * rest_size), style=f'{color} dim')
    print(f'{before}{maked}{rest}{after}')


climy/manager.py
from climy.option import Option


class Manager:
    def __init__(
            self,
            name: str,
            description: str,
            title: str = ''
    ):
        self.commands: dict[str, Manager] = {}
        self.options: dict[str, Option] = {}
        self.options_short: dict[str, str] = {}
        self.name = name
        self.title = title
        self.description = description
        self.help_item_tabsize = 25
        self.args = ''
        self.vars = []

    def add_command(
            self,
            command: 'Manager',
            *,
            name: str = ''
    ):
        name = name or command.name
        self.commands[name] = command

    def add_option(
            self,
            name: str,
            description: str,
            *,
            short: str = '',
            var_type: str = 'bool',
            var_name: str = ''
    ):
        option = Option(name=name, description=description, short=short)
        option.var_name = var_name
        option.var_type = var_type
        self.options[name] = option
        if short:
            self.options_short[short] = name

    def run(self):
        args = ''
        if not args or args[0] in ['-h', '--help', 'help']:
            text = self.generate_help()
            print(text)
            return

    def generate_help(self):
        text = ''
        text += "\r\n\r\n\033[1mUsage:\033[0m\r\n  {name} <command> [options] [arguments]".format(name=self.name)
        text += self.generate_help_options_session()
        text += "\r\n\r\n\033[1mCommands:\033[0m\r\n  {opt}".format(opt=self.generate_help_commands_session())
        return text

    @staticmethod
    def generate_help_session(name, content):
        text = "\r\n\r\n\033[1m{name}:\033[0m{content}".format(name=name, content=content)
        return text

    def generate_help_options_session(self):
        key = '-h, --help'
        desc = 'Print this help.'
        content = '\r\n  \033[36m{key: <25}\033[0m {desc}'.format(key=key, desc=desc)
        for option in self.options.values():
            key = f'-{option.short}, --{option.name}' if option.short else f'--{option.name}'
            if option.var_type != 'bool':
                var_name = option.var_name or option.name
                key += f'=[{var_name.upper()}]'
            item = '\r\n  \033[36m{key: <25}\033[0m {desc}'.format(key=key, desc=option.description)
            content += item
        text = self.generate_help_session('Options', content=content)
        return text

    def generate_help_commands_session(self):
        size = self.help_item_tabsize
        text = '\r\n  '.join([f'\033[36m{n.ljust(size)}\033[0m {c.description}' for n, c in self.commands.items()])
        return text


climy/command.py
from climy.option import Option


class Command:
    def __init__(
            self,
            name: str,
            description: str,
            title: str = '',
            handler: callable = None
    ):
        self.commands: dict[str, Command] = {}
        self.options: dict[str, Option] = {}
        self.options_short: dict[str, str] = {}
        self.options_required: set[str] = set()
        self.name = name
        self.title = title
        self.description = description
        self.help_item_tabsize = 25
        self.args = ''
        self.vars = []
        self.handler = handler
        self.parent = None

    def __str__(self):
        return f'{self.__class__.__name__}:{self.name}'

    def __repr__(self):
        return f'{self.__class__.__name__}:{self.name}'

    def add_command(
            self,
            command: 'Manager',
            *,
            name: str = ''
    ):
        command.parent = self
        name = name or command.name
        self.commands[name] = command

    def add_option(
            self,
            name: str,
            description: str,
            *,
            short: str = '',
            var_type: str = 'bool',
            var_name: str = '',
            required: bool = False
    ):
        option = Option(name=name, description=description, short=short)
        option.var_name = var_name
        option.var_type = var_type
        option.required = required
        self.options[name] = option
        if short:
            self.options_short[short] = name
        if required:
            self.options_required.add(name)

    def run(self):
        if (not self.args and not self.handler) or (self.args and self.args[0] in ['-h', '--help', 'help']):
            text = self.generate_help()
            print(text)
            return
        if self.args:
            command = self.commands.get(self.args[0], None)
            if command:
                command.args = self.args[1:]
                command.run()
                return
        for arg in self.args:
            pair = arg.split('=')
            idx = 0
            if pair[0].startswith('--'):
                idx = 2
            elif pair[0].startswith('-'):
                idx = 1
            key = pair[0][idx:]
            name = self.options_short.get(key, key)
            option = self.options.get(name, None)
            if not option:
                self.vars = name
                continue
            if name in self.options_required:
                self.options_required.remove(name)
            option.value = pair.pop(1)
        if self.options_required:
            print(f'\033[1;35mSome options are required: [{", ".join(self.options_required)}]')
            return
        if self.handler:
            self.handler(options=self.options, values=self.vars, comm=self)

    def generate_help(self, text: str = ''):
        text += "\033[1mUsage:\033[0m\r\n  {name} <command> [options] [arguments]".format(name=self.get_name())
        if self.options:
            text += self.generate_help_options_session()
        if self.commands:
            text += "\r\n\r\n\033[1mCommands:\033[0m\r\n  {opt}".format(opt=self.generate_help_commands_session())
        return text

    @staticmethod
    def generate_help_session(name, content):
        text = "\r\n\r\n\033[1m{name}:\033[0m{content}".format(name=name, content=content)
        return text

    def generate_help_options_session(self):
        key = '-h, --help'
        desc = 'Print this help.'
        content = '\r\n  \033[36m{key: <25}\033[0m {desc}'.format(key=key, desc=desc)
        for option in self.options.values():
            key = f'-{option.short}, --{option.name}' if option.short else f'--{option.name}'
            if option.var_type != 'bool':
                var_name = option.var_name or option.name
                key += f'=[{var_name.upper()}]'
            item = '\r\n  \033[36m{key: <25}\033[0m {desc}'.format(key=key, desc=option.description)
            content += item
        text = self.generate_help_session('Options', content=content)
        return text

    def generate_help_commands_session(self):
        size = self.help_item_tabsize
        text = '\r\n  '.join([f'\033[36m{n.ljust(size)}\033[0m {c.description}' for n, c in self.commands.items()])
        return text

    def get_name(self):
        if self.parent:
            return f'{self.parent.get_name()} {self.name}'
        return self.name
